Rewrite inspec-habitat plugin

jerryaldrichiii opened the pull request:
This closes #3786

This does the following to/for the `inspec-habitat` plugin:

## General Changes  
  - Adds a README.md
  - Makes default help messages clearer
  - Rewrites plugin to be more readable (at least for me)
  - Moves HEREDOCs to ERB templates
  - Set default PATH to current dir if no PATH is specified for create/setup commands
  - Adds detection of Habitat CLI config TOML in both home and `/hab/etc`

## Habitat Related

### Plan.sh:
  - Adds support for fetching profiles from Automate during build
  - Utilizes `INSPEC_CONFIG_DIR` in addition to HOME
    - HOME is still required due to `inspec/shell.rb` loading pry
  - Sets HOME and INSPEC_CONFIG_DIR via `set_runtime_env` during build
    - This was previously done via the run hook

### Run hook:
  - Adds support for creating a log file in the running service logs by default
  - Adds support for CLI reporter in supervisor logs (viewable via journalctl)
  - Adds support for reporting to Automate (configured via runtime configuration)
  - Add config options to InSpec exec (current only does `--reporter=json`)
    - Adds `target_id` for Automate reporting in Runtime

## Testing:
  - Adds a Rakefile
  - Adds integration testing for:
    - `inspec habitat profile create` (which runs `setup`)
  - Adds functional testing for `inspec habitat profile setup`
  - Adds unit testing for:
    - `inspec habitat profile create`
    - `inspec habitat profile setup`
    - `inspec habitat profile upload`

## Possible Breaking Changes:
  - Archived profile in HART moved from `$pkg_prefix/profiles/` to `$pkg_prefix`
  - `inspec habitat create/upload` no longer creates `inspec.lock` or `vendor/`
    - Does vendor in cache instead

jerryaldrichiii committed:
Rewrite inspec-habitat plugin

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

clintoncwolfe commented on lib/plugins/inspec-habitat/README.md:
```diff
@@ -0,0 +1,149 @@
+
+When running as a service, an InSpec report will be created in JSON format (by
+default at `/hab/svc/YOUR_SERVICE/logs/inspec_last_run.json`).  Additionally, a
+log of the last run can be located at
```
s/can/will/ to match previous sentence?

clintoncwolfe commented on lib/plugins/inspec-habitat/README.md:
````diff
@@ -0,0 +1,149 @@
+```
+
+This will create the following files:
+  - habitat/plan.sh (Instructs Habitat on how to build the package)
````
Ahh, @kagarmoe 	will know best, but "on how" reads colloqiual to me.  Rephrase?

jerryaldrichiii replied:
I'll change to `Provides build time instructions to Habitat`. Thoughts?

clintoncwolfe commented on lib/plugins/inspec-habitat/README.md:
````diff
@@ -0,0 +1,149 @@
+```
+
+This command will:
+  - Create a Habitat artifact (`.hart` file).
````
Does this always make a hart file?  Or only if one doesn't exist already?

jerryaldrichiii replied:
Will always create one since they are timestamped:

```
jerryaldrichiii-inspec-profile-hab_example-0.1.0-20190221211035-x86_64-linux.hart
```

clintoncwolfe commented on lib/plugins/inspec-habitat/README.md:
```diff
@@ -0,0 +1,149 @@
+
+## Testing
+
+To execute the lint/unit/functional tests run:
```
Might want to indicate the current directory

clintoncwolfe commented on lib/plugins/inspec-habitat/Rakefile:
```diff
@@ -0,0 +1,34 @@
+  Rake::TestTask.new(:unit) do |t|
+    t.description = 'Run all unit tests'
+    t.test_files = FileList['test/unit/**/*_test.rb']
+    t.warning = false
```
FWIW, I've been disabling `t.warning` - it is rarely useful and always very noisy, due to method redefs in dependent libraries.

jerryaldrichiii replied:
Went ahead and remove the file, but noted!

clintoncwolfe commented on lib/plugins/inspec-habitat/templates/habitat/hooks/run.erb:
```diff
@@ -0,0 +1,32 @@
+
+exec 2>&1
+
+CONFIG="{{pkg.svc_config_path}}/inspec_exec_config.json"
```
Perhaps I am dense, but is `{{ ... }}` ERB syntax?

jerryaldrichiii replied:
Nope, in this case it is Handlebar and is rendered in Habitat during build.

tyler-ball commented on lib/plugins/inspec-habitat/README.md:
```diff
@@ -0,0 +1,149 @@
+allows you to execute that profile as a service (via a Habitat Supervisor) on
+any Linux based platform.
+
+When running as a service, an InSpec report will be created in JSON format (by
```
These two paths look the same - are they supposed to be different since it was called out twice?

jerryaldrichiii replied:
Nope, should be different. Good catch.

tyler-ball commented on lib/plugins/inspec-habitat/README.md:
```diff
@@ -0,0 +1,149 @@
+
+### Fetching Profiles from Chef Automate During Build
+
+Chef Automate requires authentication to fetch profiles from.
```
`fetch profiles from...` where?

jerryaldrichiii replied:
Heheh, my southern English can cause problems.

"Where'd ya get the eggs from?"

I'll fix it.

clintoncwolfe commented on lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:
```diff
@@ -1,79 +1,96 @@
-
       def initialize(path, options = {})
-        @path    = path
+        @path = path
```
Very minor thing, but aligning on equal signs makes for easier to read code - here, rubocop has moved it to always be one space on each side. Wah, wah :-)

jerryaldrichiii replied:
I'm feeling rebellious! I'll align them here 😄

clintoncwolfe reviewed (commented):
Just some docs comments so far, have not read over code in detail yet.

jerryaldrichiii commented on lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:
```diff
@@ -1,79 +1,96 @@
 
-        log_level = options.fetch('log_level', 'info')
+        log_level = options.fetch(:log_level, 'info')
         @log = Inspec::Log
```
Make this an accessor

jerryaldrichiii commented on lib/plugins/inspec-habitat/Rakefile:
```diff
@@ -0,0 +1,34 @@
+require 'rake/testtask'
```
Make sure these tasks are ran in InSpec core, but leave integration test for ease of use.

jerryaldrichiii replied:
I elected to just remove all together and document the kitchen command.

tyler-ball commented on lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:
```diff
@@ -1,79 +1,96 @@
+        end
+
+        # Run create command to create habitat artifact
+        hart = create
```
If the `create` function is only every called from within this class, could we rename it to `create_hart`?

For public methods, something like `create` can make sense because when called on a class it shows purpose (`habitat_profile.create`). But for private or internally used methods I like more verbosity to easily understand purpose.

Style comment though - feel free to disagree

jerryaldrichiii replied:
It's called here https://github.com/inspec/inspec/blob/master/lib/plugins/inspec-habitat/lib/inspec-habitat/cli.rb#L17

tyler-ball commented on lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:
```diff
@@ -82,179 +99,146 @@ def upload
+      end
+
+      def verify_profile(profile)
         @log.info('Checking to see if the profile is valid...')
```
Personal style preference here. I avoid putting positive confirmation messages at `info` level and instead prefer them at `debug` level.

IMO, `info` level logs should include data like metrics - its a level that could be parsed by tooling to interpret the health of the tool. `warn` is for alerting monitors that a service is going to fail if someone does not intervene soon and `debug` (or `trace`) is for trying to trace through a program. So I would think this is a `debug` level message - it is communicating that the tool is functioning as intended, and if something goes wrong the `exit_with_error` will already communicate the issue to the user.

jerryaldrichiii replied:
Nope, seems valid to me. I'll make that change.

jerryaldrichiii replied:
I agree, should be fixed now.

tyler-ball reviewed (commented):
Had a few minor things I saw but overall it looks good. I found the README informative

jerryaldrichiii committed:
Change log message type to debug

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Make `@log` and attr_reader and rename to `logger`

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Align `=` in `initialize`

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Fix typos/grammer/etc in README.md

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Remove Rakefile (testing should be done in InSpec)

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

clintoncwolfe commented on lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:
```diff
@@ -1,260 +1,243 @@
-        log_level = options.fetch('log_level', 'info')
-        @log = Inspec::Log
-        @log.level(log_level.to_sym)
+        @logger  = Inspec::Log
```
inspec-init uses Inspec::UI to provide user feedback.  I think it would be better if inspec-habitat did the same.  That said, we could add that as a subtask on #3715 and take this as-is; it just feels bad to re-write a plugin and not use the UI system.

jerryaldrichiii replied:
I agree! Let's get this in as-is given the time constraint of me changing teams. I see you've already updated the issue on #3715, but let me know if there is anything else I can do.

clintoncwolfe commented on lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:
```diff
@@ -1,260 +1,243 @@
-
-        # returns the path to the .hart file in the work directory
-        build_hart
+        logger.debug("Creating a Habitat artifact for '#{@path}'")
```
I get a stacktrace when I don't have a hab token set:

```
[cwolfe@lodi inspec-review]$ be inspec habitat profile create hab-profile/
[2019-02-22T10:47:50-05:00] INFO: Creating a Habitat artifact for 'hab-profile/'
[2019-02-22T10:47:50-05:00] INFO: Checking to see if Habitat is installed...
[2019-02-22T10:47:50-05:00] ERROR: Unable to determine Habitat origin name.
[2019-02-22T10:47:50-05:00] ERROR: Run `hab setup` or set the HAB_ORIGIN environment variable.
[2019-02-22T10:47:50-05:00] ERROR: Unable to create Habitat artifact.
[2019-02-22T10:47:50-05:00] ERROR: RuntimeError --
bundler: failed to load command: inspec (/Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/bin/inspec)
RuntimeError:
  /Users/cwolfe/sandbox/inspec/inspec-review/lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:273:in `exit_with_error'
  /Users/cwolfe/sandbox/inspec/inspec-review/lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:186:in `verify_habitat_setup'
  /Users/cwolfe/sandbox/inspec/inspec-review/lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:31:in `create'
  /Users/cwolfe/sandbox/inspec/inspec-review/lib/plugins/inspec-habitat/lib/inspec-habitat/cli.rb:17:in `create'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
```

clintoncwolfe replied:
Ahh nuts, I forgot about this issue.

jerryaldrichiii committed:
Remove stacktrace being sent to `exit_with_error`

It is already available at the debug log level.

Thanks @clintoncwolfe for the suggestion.

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

clintoncwolfe reviewed (approved):
:ship: it!

clintoncwolfe reviewed (changes requested):
Need to fix the stacktrace mentioned on comment.

clintoncwolfe commented on lib/plugins/inspec-habitat/README.md:
````diff
@@ -0,0 +1,150 @@
+Run the following command:
+
+```
+inspec habitat profile setup PATH
````
When I run this, I see no output at all.  Is that intentional?

jerryaldrichiii replied:
Hmm...I think it was. https://github.com/inspec/inspec/pull/3818/commits/27eab8686ffda6645ff2a65e23e661abe5ac1184

I see your point though. I'll change it to output at least 1 message.

clintoncwolfe commented on lib/plugins/inspec-habitat/README.md:
````diff
@@ -0,0 +1,150 @@
+Run the following command:
+
+```
+inspec habitat profile create PATH
````
With no ENV variables set, this still throws a stacktrace:

```
[cwolfe@lodi inspec-review]$ tree my-hab-profile/
my-hab-profile/
├── README.md
├── controls
│   └── example.rb
├── habitat
│   ├── config
│   │   └── inspec_exec_config.json
│   ├── default.toml
│   ├── hooks
│   │   └── run
│   └── plan.sh
├── inspec.yml
└── libraries

5 directories, 7 files
[cwolfe@lodi inspec-review]$ be inspec habitat profile create my-hab-profile
[2019-02-28T09:16:19-05:00] ERROR: Unable to determine Habitat origin name.
[2019-02-28T09:16:19-05:00] ERROR: Run `hab setup` or set the HAB_ORIGIN environment variable.
[2019-02-28T09:16:19-05:00] ERROR: Unable to create Habitat artifact.
bundler: failed to load command: inspec (/Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/bin/inspec)
RuntimeError:
  /Users/cwolfe/sandbox/inspec/inspec-review/lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:266:in `exit_with_error'
  /Users/cwolfe/sandbox/inspec/inspec-review/lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:179:in `verify_habitat_setup'
  /Users/cwolfe/sandbox/inspec/inspec-review/lib/plugins/inspec-habitat/lib/inspec-habitat/profile.rb:30:in `create'
  /Users/cwolfe/sandbox/inspec/inspec-review/lib/plugins/inspec-habitat/lib/inspec-habitat/cli.rb:17:in `create'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:115:in `invoke'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor.rb:238:in `block in subcommand'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:115:in `invoke'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor.rb:238:in `block in subcommand'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
  /Users/cwolfe/sandbox/inspec/inspec-review/bin/inspec:12:in `<top (required)>'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/bin/inspec:23:in `load'
  /Users/cwolfe/.rbenv/versions/2.4.5/lib/ruby/gems/2.4.0/bin/inspec:23:in `<top (required)>'
[cwolfe@lodi inspec-review]$ git log --oneline -n1; date
b533958a (HEAD -> ja/rewrite-habitat-plugin, origin/ja/rewrite-habitat-plugin) Remove stacktrace being sent to `exit_with_error`
Thu Feb 28 09:18:17 EST 2019
[cwolfe@lodi inspec-review]$
```

jerryaldrichiii replied:
That stack trace is coming from Bundler. Not the InSpec process.

I'm not sure I can suppress that. Unless you know of a way?

jerryaldrichiii replied:
Nope, I'm an idiot! I called `raise` without handling any exceptions. You get a different stack trace when not using Bundler, but a stack trace is a stack trace.

clintoncwolfe reviewed (dismissed):
I still see a stacktrace :-(

jerryaldrichiii committed:
Modify output to be a bit better for the user

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Prevent stack trace from being set to the user

This also verifies Habitat setup during create as well

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

clintoncwolfe reviewed (approved):
There are still some rough edges, but the need was to get the plugin working, and this PR does that.  I noticed a UI issue, opened #3877 to cover it later.

Thanks @jerryaldrichiii!
